src/main/java/com/limechain/storage/block/BlockState.java

```diff
@@ -904,4 +918,56 @@ public void handleFinalizedBlock(final Hash256 currentFinalizedHash) {
             //TODO: If currentFinalizedHash is not equal to subchain hash, delete subchain state trie
         }
     }
+
+    public synchronized void addBlockToBlockTree(BlockHeader blockHeader) {
+        if (isInitialized()) {
+            if (isFullSyncFinished()) {
+
+                processPendingBlocksFromQueue();
```

Comment:
Is there a reason why we try to reprocess the queue every time we add instead of processing once after full sync finish?

Reply:
With the current implementation, there should be only one invocation of the method, but for my previous attempts, it was necessary to process them at least twice and I missed changing that.

Reply:
After some investigation, I recalled the exact reason I chose to do it every time. Making processPendingBlocksFromQueue() public and calling it from FullStateMachine introduces the risk of adding new blocks from block announcements while the old ones are being processed. Since I'm using a non-thread-safe queue, this can lead to unpredictable behavior. By making only the addBlockToBlockTree public and synchronized at the same time, I can be sure that new blocks will not be added while the old ones are processed. With that in mind, there are two options: 1. To keep that solution (which imo doesn't have big overhead because if the queue is empty the processPendingBlocksFromQueue() does nothing) or 2. To use a thread-safe queue. @Zurcusa What do you think would be the better approach?

Reply:
You're right. Overhead is close to none. No need to waste time changing it.